pkg/archive/archive.go

```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		}
 
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```

Comment:
> ... that ship sailed 3 years ago

It's only been 2 years! https://github.com/moby/moby/pull/13171 🎉 

Also, since the `docker cp` logic that does the name fudging is done entirely in the CLI, we should be able to redesign it without affecting older clients.

One alternative that I have warmed up to is that if the source of the archive is a directory, then create the archive by going into that source directory and archiving `.` such that all archive entries begin with `.` and then the extract logic would be:

1. Create the destination directory if it does not exist.
2. Extract the archive in the destination directory.

That should handle the case where the name of the destination directory is not the same as the source directory. That would do away with the need to change the archive entries on the fly.